spec/ics-3-bidirectional-connection-semantics/README.md

```diff
@@ -0,0 +1,358 @@
+
+This subprotocol will likely need to be permissioned to an entity who "owns" the connection on the initiating chain, such as a particular end user, smart contract, or governance mechanism.
+
+The closing handshake subprotocol defines three datagrams: *ConnCloseInit*, *ConnCloseTry*, and *ConnCloseAck*.
+
+A correct protocol execution flows as follows:
+
+| Initiator | Datagram          | Chain |
+| --------- | ----------------- | ----- |
+| Actor     | `ConnCloseInit`   | A     |
+| Relayer   | `ConnCloseTry`    | B     |
+| Relayer   | `ConnCloseAck`    | A     |
+
+*ConnCloseInit* initializes a close attempt on chain A.
+
+```golang
+type ConnCloseInit struct {
+  Identifier identifier
+  Identifier identifierCounterparty
+}
+```
+
+```coffeescript
+function handleConnCloseInit(identifier, identifierCounterparty)
+  (state, version, counterpartyIdentifier, rootOfTrust) = Get(identifier)
+  assert(state == OPEN)
+  assert(identifierCounterparty == counterpartyIdentifier)
+  state = TRYCLOSE
+  Set(identifier, (state, version, counterpartyIdentifier, rootOfTrust))
+```
+
+*ConnCloseTry* relays the intent to close a connection from chain A to chain B.
+
+```golang
+type ConnCloseTry struct {
+  Identifier identifier
+  Identifier identifierCounterparty
+  AccumulatorProof proofInit
+}
+```
+
+```coffeescript
+function handleConnCloseTry(identifier, identifierCounterparty, proofInit)
+  (state, version, counterpartyIdentifier, rootOfTrust) = Get(identifier)
+  assert(state == OPEN)
+  assert(identifierCounterparty == counterpartyIdentifier)
+  assert(verify(roofOfTrust, proofInit, (counterpartyIdentifier, TRYCLOSE)))
+  state = CLOSED
+  Set(identifier, (state, version, counterpartyIdentifier, rootOfTrust))
+```
+
+*ConnCloseAck* acknowledges a connection closure on chain B.
+
+```golang
+type ConnCloseAck struct {
+  Identifier identifier
+  AccumulatorProof proofTry
+}
+```
+
+```coffeescript
+function handleConnCloseAck(identifier, proofTry)
+  (state, version, counterpartyIdentifier, rootOfTrust) = Get(identifier)
+  assert(state == TRYCLOSE)
+  assert(verify(rootOfTrust, proofTry, (counterpartyIdentifier, CLOSED)))
+  state = CLOSED
+  Set(identifier, (state, version, counterpartyIdentifier, rootOfTrust))
+```
+
+##### Closing by Equivocation
+
+The equivocation closing subprotocol serves to immediately close a connection if a consensus equivocation is discovered and thus prevent further packet transmission.
+
+The equivocation closing subprotocol defines only one datagram, *ConnCloseEquivocation*:
+
+```golang
+type ConnCloseEquivocation struct {
+  Identifier identifier
+  Header firstHeader
+  Header secondHeader
+}
+```
+
+```coffeescript
+function handleConnCloseEquivocation(identifier, firstHeader, secondHeader)
+  (state, version, counterpartyIdentifier, rootOfTrust) = Get(identifier)
+  assert(state == OPEN)
+  assert(checkEquivocation(rootOfTrust, firstHeader, secondHeader))
```

Comment:
Shouldn't this take the current trust state not the root of trust?

Reply:
It is that, I guess the nomenclature is confusing. I think we will change it to `ConsensusState`.

edit: renamed